### ISSUE TITLE
docker cp into a sysbox container results in files with nobody:nogroup.

### ISSUE THREAD

**Issue by samuelba** [nestybox/sysbox#866]
**Version**: v0.6.5 (same for v0.6.4)

**Host**: Ubuntu 24.04

Installed sysbox on host with the provided debian package from the release https://github.com/nestybox/sysbox/releases/tag/v0.6.5

Run

```
sudo docker run --runtime=sysbox-runc --rm -it --hostname my_cont --name my_cont registry.nestybox.com/nestybox/ubuntu-focal-docker bash
```

Then copy a file from host to the container e.g. 

```
sudo docker cp some_file.txt my_cont:/
```

Then inside the container do `ls -lah /`, you will see the file owner (user/group) is `nobody`

```
-rwxrwxr-x   1 nobody nogroup 2.1K Oct 28 19:24 some_file.txt
```

Then try to set root as the owner

```
chown root:root /some_file.txt
```

This will return this error 

```
chown: changing ownership of 'some_file.txt': Operation not permitted
```

The same steps work on host Ubuntu 20.04. Also, after `docker cp` the file to the container, the file actually is owned by `root` and not `nobody`.

**Comment by ctalledo** [comment]
Hi @samuelba, apologies for the late reply.

The problem you see occurs because Sysbox creates containers with the Linux user-namespaces, meaning root the user in the container maps to an unprivileged user on the host. `docker cp` is currently not aware of this, so when it copies the file from the container to the host, it does not remap the file ownership. This is why you see the copied files show up with `nobody:nogroup` ownership inside the container.

As a work-around, there's a tool called [sysbox-docker-cp](https://github.com/nestybox/sysbox/blob/master/tests/scr/sysbox-docker-cp) that does the remapping of file ownership. It's a simple bash script wrapper around `docker cp`. You can give that a try. It does need `sudo` permissions when copying files from the host to the container, because it needs to access the container's root-filesystem on the host which is root-only access under `/var/lib/docker/...`. 

Hope that helps.

**Comment by ctalledo** [comment]
Closing based on previous answer; please re-open if needed.

**Comment by ctalledo** [comment]
Hi @samuelba, 

> Do you know why that worked on 20.04 but not anymore on 24.04? Did something change on 24.04 that docker cp does not work anymore?

Not 100% sure, but I suspect it has to do with kernel support for overlayfs on idmapped mounts (i.e., it's landed in kernel 5.19). When available, Sysbox uses it instead of shiftfs. It's better, but unfortunately breaks `docker cp` (which I had not realized until you opened this issue).

**Comment by samuelba** [comment]
Thank you, that would make sense.

Anyway, `sysbox-docker-cp` works.
